### ISSUE TITLE
Add support for the d4 format?

### ISSUE THREAD

**Issue by tfenne** [igvteam/igv#1027]
Hi @jrobinso - we've been talking with @arq5x about wanting to add support for [d4](https://github.com/38/d4-format) to IGV.  It sounds like there is some prior work and/or discussion on this but I'm having trouble finding it.  From chatting to @arq5x it sounds like you didn't like the approach taken - and it sounds like perhaps it was due to bundling in of native code within the IGV repo?

Before we start work on it I wanted to propose a different route and see if you would be on board.  My thinking is that we'd create a d4-java repo that would:

- Implement a set of Java classes for reading and querying d4 format
- Almost certainly that library would wrapper the current native implementation using JNI of JNA
- Bundle the shared library file for d4 inside the JAR that is published to Maven (the way that e.g. snappy and the intel gzip stuff is bundled internally in HTSJDK)
- Have the Java API implement the appropriate tribble interfaces

Then once that's stabilized we'd make a PR into IGV to add support for D4 using that library.

Is this an approach you'd support and be willing to accept a PR into IGV for once the external d4-java library exists?

**Comment by arq5x** [comment]
Thanks for launching a discussion, @tfenne. My memory of the conversation we had with @jrobinso over email is that he was hesitant about the fact that @38's fork of IGV (https://github.com/38/igv) used native Rust code for the d4 support.  As there is growing interest in using D4 for the IGVF consortium, we are very eager to build support for D4 in IGV, other browsers, and R.

**Comment by jrobinso** [comment]
@tfenne How would distribution of native libraries work?  Would we need to maintain separate builds of IGV for every platform, including multiple Linux platforms?  I've been through this in ages past with HDF5 and am not keen for a repeat.  I don't really know how snappy works but I don't think you distribute jars for the htsjdk, only source code.

Would it be possible to design this so the jar is downloaded and loading dynamically by users that are interested in it?

**Comment by jrobinso** [comment]
@arq5x BTW I have no objection for you or anyone else to distribute a D4 aware IGV, we could even link to it from our downloads page.   I won't do that to you however without your blessing.

For really wide spread support in genome browsers I think you need native readers for Java and Javascript,  how difficult would that be?  Naively I see that as just a language translation task.

**Comment by maximilianh** [comment]
Ideally also a C library. Some people still use this language. :-)

**Comment by arq5x** [comment]
@maximilianh - ha! @38 has a C-API to the D4 readers and writers if interested: https://github.com/38/d4-format/blob/master/d4binding/include/d4.h

@jrobinso Thanks for letting us know. Ideally, we could find a workable solution to provide D4 support in the main IGV, but this is always an alternative.

**Comment by tfenne** [comment]
@jrobinso I started writing a response somewhere and the lost it.  Basically we'd need to figure out all the platforms that IGV supports (I know in theory, being Java the answer is largely "all of them", but realistically), and then we would a) cross-build the rust library and generate a library per platform (or group of platforms).  Those would get stored in the JAR file that is uploaded to maven central, and when the Java API is used, the relevant library would be unpacked into a temporary directory and loaded as a native library by the JVM process.  That's how we do it for several native libraries across HTSJDK/Picard/GATK.

That said, I can understand not necessarily wanting the support headache that comes from users complaining to _you_ about why D4 support doesn't work on _insert configuration here_.

We can certainly think about what it would take to at least provider readers for D4 in pure Java.  I _think_ part of the concern (and @38 or @arq5x can comment better) is that it is still a young format and evolving and so wrappering the standard implementation seems safer than having a separate Java implementation that needs to evolve in parallel.

**Comment by jrobinso** [comment]
@tfenne Its amazing how many platform variants there are, its the reason hdf5 support was abandoned.     Speaking if HTSJDK,  doesn't this really belong there?

**Comment by 38** [comment]
> Ideally also a C library. Some people still use this language. :-)

I know your concern, the fact is Rust has a much cross-platform friendly tool chain than C/C++. And I believe the language doesn't seem to cause any problem.



> For really wide spread support in genome browsers I think you need native readers for Java and Javascript, how difficult would that be? Naively I see that as just a language translation task.

Just FYI, the point of having D4 format is to fully utilize the hardware. It's definitely possible to write a paser in Java or Javascript, however, the performance would most likely be extremely slow which loses the point of the format. What a D4 format can bring is a new ability that you can even compute your customized per-base score and explorer it using IGV interactively. But a Java/Javascript implement doesn't seem to be very useful. 



> @tfenne How would distribution of native libraries work? Would we need to maintain separate builds of IGV for every platform, including multiple Linux platforms? I've been through this in ages past with HDF5 and am not keen for a repeat. I don't really know how snappy works but I don't think they distribute jars for the htsjdk, only source code.

I totally understand the pain of maintaining a cross-platform JNI package. So I guess we would like take the responsibility of maintaining it if IGV has a plugin system, so that we can maintain our own IGV-D4 plugin on our side. Please let me know if IGV currently support this. 



> a) cross-build the rust library and generate a library per platform (or group of platforms). Those would get stored in the JAR file that is uploaded to maven central, and when the Java API is used, the relevant library would be unpacked into a temporary directory and loaded as a native library by the JVM process. That's how we do it for several native libraries across HTSJDK/Picard/GATK

And @tfenne , I think that could be one possible solution. But I guess the IGV folk still doesn't convinced this won't cause any compatibility issue. As far as I can see the problem is a little bit trick since the JNI binary is also coupled with the Java runtime version, this means we probably need to compile multiple binary for each JRE implementation (as far as I awared of JNI binary for Sun JDK and OpenJDK is not compatible, but I might be wrong).

**Comment by jrobinso** [comment]
@38 we do not have a general plug-in system but we could conceivably do something ad-hoc for this.   I would probably need someone else to do the heavy lifting here and make a PR.   

I really think the proper project for this is the HTSJDK,  surely IGV isn't the only Java based program that would use these files.   What of the GATK?

I don't understand the performance concern,  IGV already computes per-base coverage on-the-fly.    It is primarily, of course, a visualization package.   I don't see a use case now where performance would be a consideration for IGV, but perhaps you can elaborate one.

**Comment by 38** [comment]
> I don't understand the performance concern, IGV already computes per-base coverage on-the-fly.

D4 actually enables realtime value aggregation even in the chromosome level (the typical running time for compute the mean depth with D4 format for Chr1 is a few hundred ms) - unlike IGV does, D4 will enable real-time visualization even you are viewing a very large range. Also, it's quite different from BW files as well, since we are actually compute the fresh values and aggregate it, which allows you aggregate the value differently. 

I guess it's definitely useful - for example, IGV's builtin depth visualization doesn't work when you are viewing a large region.

**Comment by 38** [comment]
> I guess it's definitely useful - for example, IGV's builtin depth visualization doesn't work when you are viewing a large region

In addition to that, BigWig doesn't solve the problem either, as profiling a bam to a BW file may spend hours. But if D4 was introduced, profiling a bam to D4 only takes a few minutes. And in addition to that, you can view the depth in arbitrary range efficiently. As a IGV user, it's always annoying me when a SV is very large, the IGV is constantly loading. I think this is definitely a use case.

**Comment by jrobinso** [comment]
What is the current "market share" of D4 files vs bam/cram?

**Comment by jrobinso** [comment]
Another question, this only works I assume if files are local to a user's machine,  a large % of IGV usage now is loading through the cloud (by https or htsget).

**Comment by tfenne** [comment]
Re: market share - I don't know, but I think the question is more D4 vs. BigWig and other genome-wide count formats rather than SAM/BAM.

**Comment by arq5x** [comment]
Yes, D4 is an alternative to BigWig not BAM/CRAM. Our desire is to find a solution that allows users to load D4 files of quantitative genomes data into IGV. It would be great to settle on a solution that works for both the IGV and D4 development teams.

**Comment by jrobinso** [comment]
@tfenne That sounds reasonable,  my object is mainly to building and distributing machine dependent binaries.   I'm having a hard time seeing how this works in practice but I haven't dealt with JNI in a long time.

@38 's use case was computing coverage on-the-fly from a BAM file,  at least that's how I read it,  that's why my focus on BAM.   If this is really just a counting format the htsget idea could still apply.   The idea would be to define a simple http get based protocol for a "server",  then implement servers (in C perhaps) for various platforms.   The server could either run as a true server,  or on the users local machine with a tiny web server.   The server only has to understand http get.   As you know IGV itself is such a server, listening on port 60151.   Let me know your thoughts on this, not only would it solve the local file probably but double as a cloud solution, and be immediately useful for igv.js which right now has ~100,000 launches per week, a few thousand more than desktop.

**Comment by 38** [comment]
> As you know IGV itself is such a server, listening on port 60151. Let me know your thoughts on this

Oh, that's the thing I am not aware of. And is there any documentation for the protocol and how can we load a remote server ?

**Comment by jrobinso** [comment]
@38 I'm not sure i understand your question, but see the IGV documentation for communicating to IGV via port of http get commands.   Its not immediately relevant here, just pointing out that a simple local http "server" can be implemented in a few lines of code.   The idea would be to wrap a D4 reader with such a server, and IGV would communicate with it, not the other way around.

Of course IGV loads data already by URL,  is that what you are asking?   

No htsget does not,  its just an exemplar, this would be new.   It would take 1/2 a day to write such a server.   This is how IGV served BAM files over http originally,  before I became aware of "range" requests.

**Comment by jrobinso** [comment]
I envsion something like this

http://myD4Server/pathToMyFile?chr=1&start=100&end=10000

returns stream of coverage/count data

**Comment by 38** [comment]
> I envsion something like this
> 
> http://myD4Server/pathToMyFile?chr=1&start=100&end=10000
> 
> returns stream of coverage/count data

Of course, is this already supported in IGV? Can user load the track with URL http://myD4Server/pathToMyFile ?

**Comment by jrobinso** [comment]
@38  Is what exactly already supported?   I'm assuming a little IGV knowledge, but I'm not sure we are communicating.    All sorts of URLs are supported, this would potentially be a new one but adding support should be pretty trivial.   We actually run a server now for users who's local network does not allow "range" requests with exactly the format above.

**Comment by jrobinso** [comment]
Or perhaps this is too clumsy,  if the majority of D4 users are going to download and load local files, and there is no interest in supporting loading over the web.    OTOH it is really easy to do.

**Comment by 38** [comment]
> majority of D4 users

I guess one of the use case would be fast profiling a local BAM file and visualize the depth in IGV. As I mentioned before, profiling BAM to D4 is very efficient and I think this is a improvement for the pipeline. 

In addition, I think accessing via HTTP server is a good idea anyway - but I am not aware of existence of any IGV supported protocol (I am not talking about HTTP protocol and I know IGV can load data through HTTP. I am asking if IGV can access quantitative data over HTTP server and let the server do the aggregation step - range request simply doesn't work, because of the network bandwidth) 

I

**Comment by jrobinso** [comment]
@38 Yes of course it can, htsget is an excellent example and is close to what I am proposing here.   I will write an example server in python for bigwig files and post it later today, this discussion is getting a bit silly.

**Comment by arq5x** [comment]
Thank you all for the discussion and thank you @jim for enlightening us to possible solutions we hadn't considered.

**Comment by 38** [comment]
>  will write an example server in python for bigwig files and post it later today

Yes, that's really helpful, thanks for the help! Looking forward for the example!

**Comment by jrobinso** [comment]
OK that took 5X longer than expected, which is better than usual.   Here you will find a prototype "D4" server, with the important caveat that it really serves bigwig files.    There is a README,  but briefly you start the server (after installing the python requirements in requirements.txt) as follows.  Requires a recent version of python, probably.

```
python d4serer.py
```

After pulling  building IGV from the "d4" branch you can load a bigwig file with a url such as

```
d4get://localhost:8888/Users/jrobinso/igv-team%20Dropbox/Data/Juicebox/CTCF_Untreated.bw
```

The path of the URL (everything after 8888) is a full path to a bigwig file on the computer the server is running. 

Something like this is what I was trying to describe,  with bigwig serving as a proxy for d4.    

Its obvious how you would use this to serve d4 files over the cloud.   Using it locally would be clumsy if the user had to paste in the URL from the file menu, so we wouldn't use it like that.   The user would install the server on their machine, and tell IGV where it lives, a bit like bedtools is handled.    They would browse to d4 files using the file menu,  IGV would detect by some means that this is a d4 file and route the request to the server.  

This is obviously a prototype for discussion purposes, in case there is any misunderstanding on that.

**Comment by jrobinso** [comment]
@tfenne This does not preclude your approach of course, they aren't mutually exclusive, and something will be needed for remote access in any event if this format becomes widely used.

**Comment by arq5x** [comment]
Thank you for putting this together, @jrobinso! I will defer to @38 and @tfenne regarding how to love forward, but I will say that I see remote file support as essential given our goal of having D4 used in the storage of datasets from the IGVF consortium. Remote file support for D4 is something they are keen on as well.  I think @38 was also considering writing a `d4 server $D4FILE` tool that could potentially respond to requests on a particular port for either local or remote files. I am a bit out of my realm here, so I will simply state our functionality goals and leave the discussion of the best approach to the experts.

**Comment by 38** [comment]
> After pulling and building IGV from the "d4" branch you can load a bigwig file with a url such as

Thanks @jrobinso , this is great. I am going to implement the server side code and try this out.

**Comment by brainstorm** [comment]
> Another question, this only works I assume if files are local to a user's machine, a large % of IGV usage now is loading through the cloud (by https or htsget). Given that htsget is supported, one solution for many might be to implement an htsget server that uses D4 on the backend.

htsget is supported on IGV, but I'm not sure which endpoint(s) would fit D4 the best? Currently htsget supports `/reads` or `/variants`, I guess D4 would pertain to both and more since it also supports BigWig/BED?

Even if it's not in the spec, it'd be relatively simple to add (provided that the Rust API exposes the necessary methods, haven't explored closely yet), @38 check our htsget Rust implementation for several other hts formats:

https://github.com/umccr/htsget-rs/tree/main/htsget-search/src/htsget

/cc @ohofmann @mlin

**Comment by ohofmann** [comment]
We've had discussions around support of other formats (BED and BigWig were mentioned) for a future release. I'll defer to @mlin - not sure what the roadmap looks like, but this could be a good opportunity.

**Comment by arq5x** [comment]
htsget support for interval and quantitative formats would be great.

**Comment by jrobinso** [comment]
Yes htsget would be good, but would probably not be a great solution for local files.

**Comment by arq5x** [comment]
What are the solutions in play for local files?  I know the server approach @jrobinso prototyped is one and the 3rd party library with small IGV PR proposed by @tfenne is another. Anything else I am missing?

**Comment by jrobinso** [comment]
@arq5x I think you covered it.

**Comment by arq5x** [comment]
Based on this helpful discussion, @38 is moving forward with implementing a `d4tools server my.d4` for use with an anticipated htsget specification for quantitative formats (assuming roughly `chr`, `start`, `end`, `value`), following the very helpful example server from @jrobinso. If this is designed to support both local and remote files, it seems that it will prevent the need for internal changes to IGV and the need for external libraries.  Please correct me if I am mistaken.  

@mlin - where should I start an issue to move forward with developing a spec for interval and and quantitative interval support in htsget?

**Comment by jrobinso** [comment]
@arq5x @38 @mlin Note that in my server implementation there are 2 methods.   1 a query to get the data for a region, and one to get the chromosome names.   Both are essential if IGV is to handle chromosome aliasing.   1 <-> chr1 is pretty easy, but those aren't the only name conventions in use.   htsget does not explicitly have such a method, we rely on "class=header" for that.   As d4 files probably don't have a header you might need to make something up.   I don't really care as long as I can get the sequence (chromosome) names.

htsget is a bit clumsy in that you first make a query for "tickets",  which are then used to fetch the actual data.   If you want to implement that spec its fine of course, just more complex than what I gave you.

**Comment by brainstorm** [comment]
> @mlin - where should I start an issue to move forward with developing a spec for interval and quantitative interval support in htsget?

Yes please @arq5x @38, the place to file that issue should be in https://github.com/samtools/hts-specs/issues, where the htsget spec lives.

**Comment by jrobinso** [comment]
@38 It occurs to me that it would be useful to pass a desired resolution (bin size) for aggregated counts as an optional parameter, in addition to chr, start, end.   This is implied in the prototype bigwig implementation,  where bin size is calculated as (end-start)/1000.    The 1000 is just a nominal screen width,  so that calculation is really for desired base pairs per bin, or stated another way bin size in base pairs.    There's some funny weird math to workaround some pyBigWig rules, please ignore that.   In any event since with d4 you can aggregate counts on the fly this should be an (optional) parameter.   Another useful parameter would be the aggregation statistic,  mean, median, min, max, percentile, or whatever you can support.   These are just some ideas.

**Comment by 38** [comment]
Thanks @jrobinso  Bin size can be an additional parameter for sure.

I also had the prototype done on my side https://github.com/38/d4-format/blob/master/d4utils/src/server/main.rs And it worked with d4 branch. 

For different stats I think it's better to have "d4get://localhost:60000/mean" or "d4get://localhost:60000/percentile_90" for example. So it doesn't needs to be inside the protocol and later we may want to introduce multi-sample aggregation as well.

And I think we would like to wait until the hts-spec gets ready, then I can move forward on a formal implementation of D4 server. 

Also, I would like to discuss the possibility of make IGV automatically spawn a D4 server when user tries to open a local file. But I am not sure if you guys have any concern regarding this. 

Thanks for the help!

**Comment by 38** [comment]
> @38, please let me know where that branch lives so that I can write it in our htsget rust server, as a pre-spec preview.

Sorry for the ignorance, but what branch are you asking for ? If you are asking the prototype of D4 server that works with @jrobinso 's prototype, please check https://github.com/38/d4-format/blob/master/d4utils/src/server/main.rs

**Comment by jrobinso** [comment]
Where did everybody go?   @38 @tfenne?

**Comment by tfenne** [comment]
And from my side, it sounds like @arq5x and @38 are happy with an HTTP solution for now, so we're not actively pushing forward on a non-HTTP solution.
